File: chapter8/cpp/chapter8.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include "chapter8.h"

template <class T>
SLNode_t<T> *sl_node(const T &value)
{
  SLNode_t<T> *node = new SLNode_t<T>(value);
  return node;
}

template <class T>
SList_t <T> *sl_list()
{
  SList_t <T> *list = new SList_t <T> ();
  return list;
}

template <class T>
void SList_t<T>::pushFront(const T &value) 
{
  SLNode_t<T> *node = sl_node<T>(value);

  if (head == NULL) {
    head = tail = node;
  } else {
    node->next = head;
    head = node;
  }
}



template <class T>
ret_obj<T> SList_t<T>::popFront()
{
  ret_obj<T> data{false};
  SLNode_t<T> *node; 
  
  if (head == NULL) 
    return data;

  data.valid = true;
  data.value = head->value;
  node = head;
  
  if (head == tail  && tail == node) 
    tail = head = NULL;
  else 
    head = head->next;

  free(node); 
  return data;
}

template <class T>
void SList_t<T>::pushBack(const T &value)
{
  SLNode_t<T> *node = sl_node<T>(value);

  if (head == NULL) {
    head = tail = node;
  } else {
    tail->next = node; 
    tail = node;
  }
}

template <class T>
ret_obj<T> SList_t<T>::popBack()
{
  ret_obj<T> data {false};
  SLNode_t<T> *runner, *runner_prev;

  if (head == tail) 
    return data; 
  
  runner = head;
  data.valid = true;

  if (head == tail) {
    head = tail = NULL;
    data.value = runner->value; 
    delete(runner);
  } else {
    while (runner->next) {
      runner_prev = runner;
      runner = runner->next;
    }
    data.valid = runner->value;  
    free(runner);
    tail = runner_prev;
    runner_prev->next = NULL;
  }

  return data;
}

template <class T>
bool SList_t<T>::removeValue(const T &value)
{
  SLNode_t<T> *runner = head, *runner_prev = NULL, *tmp_node;
  bool rslt = false; 

  while (runner) {
    if (runner->value == value) {
      tmp_node = runner;
      if (head == runner) {
        head = head->next;
        runner = head;
        runner_prev = NULL;
      } else {
        runner_prev ->next = runner->next;
        runner = runner_prev->next;
      } 

      if (tmp_node == tail) {
        tail = runner_prev;
      }
      free(tmp_node);
      rslt = true;
    } else {
      runner_prev = runner;
      runner = runner->next;
    }
  }
  return rslt;
}

template<typename T>
void SList_t<T>::reverse (SList_t<T> *list)
{
  
  SLNode_t<T> *tmp_node = sl_node<T>(0), *tmp_node_next = NULL;

  if (list == NULL) 
    list = this;
  
  if (list->head == NULL) 
    return;
  
  while (list->head) {
    tmp_node->next = list->head;
    list->head = list->head->next;
    tmp_node->next->next = tmp_node_next;
    tmp_node_next = tmp_node->next;
  }

  if (tmp_node->next) 
    list->head = tmp_node->next;
  
  free(tmp_node);
}

template <typename T>
bool SList_t<T>::isPalindrome() {
  SLNode_t<T> *runner = head, *runner_end = tail, *runner_start = head;

  if (runner_start == NULL)
    return false; 

  while (runner_start != runner_end && runner_end->next != runner_start) {
    runner = runner_start;
    
    while (runner->next != runner_end) 
      runner = runner->next;
    
    if (runner_start->value != runner_end->value )
      return false;

    runner_start = runner_start->next; /* update start */
    runner_end = runner; /* update end */
  }

  return true;
}

template <class T>
ret_obj<T> SList_t<T>::kthLastNode(SList_t *list, unsigned int k)
{
  SLNode_t<T> *runner = NULL;
  ret_obj<T> data{false};
  int count = 0;

  if (list == NULL) 
    runner = head;
  
	/* count number of nodes */
  while (runner)  {
    runner = runner->next;
    count++;
  }
  
  count -= k;
  /* point runner to kth node  */
  if (count > 0) {
    runner = head;
    while (count--) 
      runner = runner->next;
  }

  if (runner) {
    data.valid = true;
    data.value = runner->value;
  }

  return data;
}

template <class T>
void SList_t<T>::shiftRight(int n)
{
  SLNode_t<T> *runner = head;
	SLNode_t<T> *newHead = NULL, *oldHead = NULL;
  int count = 0;

  if (head == NULL)
    return; 
  
	while (runner) {
		runner = runner->next;
		count += 1;
	}

	if (n > 0 && count > 1) {
		runner = list->head;
		for (int i = 0; i < abs(n - count); i++) {
			runner = runner->next;
		}

		/* break list */
		newHead = runner->next;
		runner->next = NULL;

    runner = newHead;
    
    /* connect new head's tail to old head*/
    while (runner->next) 
      runner = runner->next;
    runner->next = head;
    head = newHead;
	}

	if (n < 0) {
		return shift_left( abs(n));
	}

}

template <class T>
void SList_t<T>::shiftLeft(int n)
{
  SLNode_t<T> *runner = head;
	SLNode_t<T> *newHead = NULL;
	int count = 0;
  
	while (runner) {
		count += 1;
		
		if (count == n) {
		
		/* break link */
			newHead = runner->next;
			runner->next = NULL;
		
		/*new Head must be NULL*/
			if (newHead == NULL) 
				break;
			
		/* point runner to tail of newHead */
			runner = newHead;
			while (runner->next) 
				runner = runner->next;
			
		/* link new list with new Head*/
			runner->next = list->head;
		
		/* update head */
			list->head = newHead;

			break;
		}
		runner = runner->next;
  }
}

template <class T>
void  SList_t<T>::display()
{
  SLNode_t<T> *runner = head;
  while (runner) {
    std::cout <<  "[" << runner->value << "]";
    runner = runner->next;
  }
}

template <class T>
SList_t<T>  *SList_t<T>::sum_numerals (SList_t<T> *list_a, SList_t<T> *list_b, bool least_sig_digit_first)
{
  SList_t<T> *newList = sl_list<T>();
  SLNode_t<T> *a = list_a->head;
  SLNode_t<T> *b = list_b->head;
  T sum[3] = {0,0,0};
  int factor  = 1;
  int count = 0;
  int temp;

  while (a || b ) {
    if (a) {
      sum[0] += a->value%10 * factor;
      a = a->next;
    }
    if (b) {
      sum[1] += b->value%10 * factor;
      b = b->next;
    }
    factor *= 10;
    count++;
  }
  sum[2] = sum[0] + sum[1];
  
  while (count--) {
    if (least_sig_digit_first) {
      newList->pushBack(sum[2] % 10);
      sum[2] /= 10;
    } else {
      factor/=10;
      newList->pushBack(sum[2] / factor);
      sum[2] -= (sum[2] / factor)*factor;
    }
  }
  newList->display();
  return newList;
}

template <class T>
SList_t<T> *SList_t<T>::setupLoop(unsigned int count, unsigned int loop_pos)
{
  SList_t<T> *circularList = sl_list<T>();
  SLNode_t<T> *ref;
  for (int i = 1; i <= count; i++) {
    circularList->pushBack(i);
    if (i == loop_pos)
      ref = circularList->tail;
  }
  tail->next = ref;
}


template <class T>
bool SList_t<T>::hasLoop()
{
  SLNode_t<T> *slow =, *fast;
  bool hasLoop = false;

  if (head == NULL)
    return hasLoop;
  slow = fast = head; 
  while (slow && !hasLoop) {
    slow = slow->next;
    for (int i = 2; i-- && fast ;)
      fast = fast.next;
    hasLoop = fast == slow;
  }
  return hasLoop;
}

template <class T>
SLNode_t<T> *SList_t<T>::loopStart()
{
  SLNode_t<T> *slow, *fast, *collision = NULL, *loopHead, *runner = NULL;
  unsigned int n = 0;

  if (head == NULL)
    return runner;

  slow = fast = head;
  
  while (slow) {
    slow = slow->next;
    
    for (int i = 2; i-- && fast ;) 
      fast = fast->next;
    
    if (fast == slow)
      collision = slow; 
  }

  /* number of nodes in loop  */
  if (collision) {
    runner = collision->next;
    while (runner != collision) {
      n++;
      runner = runner->next;
    }

    /* move loop head n-1 after LL Head */
    loopHead = head;
    for (int i = n-1; i-- ;) 
      loopHead = loopHead->next;

    runner = loopHead;
    /* set runner to feedback node */
    while (runner->next != loopHead) 
      runner = runner->next;
  }

  return runner; 
}

template <class T>
void SList_t<T>::breakLoop()
{
  SLNode_t<T> *feedback = loopStart();
  if (feedback)
    feedback->next = NULL;
}

template <class T>
unsigned int SList_t<T>::nodeCount()
{
  SLNode_t<T> *feedback = loopStart(), *node = head;
	unsigned int count = 0;

	if (node == NULL) {
    return count;
  }

	while (node->next != feedback) {
		count++;
		node = node->next;
  }

	return count;
}

template <class T>
void SList_t<T>::swapPairs()
{
  SLNode_t<T> *runner = head, *runner_prev;
  while (runner) {
    if (runner_prev) {
      if (head == runner_prev)
        head = runner;
      if (tail == runner)
        tail = runner_prev;
      
    	runner_prev->next = runner->next;
			runner->next = runner_prev;

      runner = runner_prev;
      runner_prev = NULL;

    }
    runner_prev = runner;
    runner = runner->next;
  }
}

template <class T>
void SList_t<T>::flattenChildren()
{
  std::queue<SLNode_t<T>*> collection;
  bool parent_list_done = false;
  SLNode_t<T> *runner; 
  
  if (head == NULL)
    return;

  collection.push_back(head);
  
  while (collection.size()) {
    runner = collection.front();
    collection.pop();

    if (parent_list_done) 
      tail->next = runner; /*connect previous tail to child */

    while (runner) {
      if (runner->child)
        collection.push(runner->child);

      if (runner == tail)  /* parent list traversal complete */
        parent_list_done = true;

      if (parent_list_done) /* update tail */
        tail = runner;

      runner = runner->next;
    }
  }
}

template <class T>
void SList_t<T>::unflattenChildren()
{
  SLNode_t<T> *runner = head, *runner_prev = NULL; 
  std::map<SLNode_t<T>*, unsigned int> histo;
  bool first_child_removed = false;

  if (runner == NULL)
    return;

  while (runner) {
    if (runner->child) {
      if (!histo.count(runner->child)) {
        histo.insert({runner->child, 0});
      } else {
        histo.at(runner->child)++;
      } 
      if ( (first_child_removed == false && histo.at(runner->child) == 2) || first_child_removed == true) {  /* children found after histo == 2 are unlinked*/
        runner_prev->next = NULL; 
      }
    }
    runner_prev = runner; 
    runner = runner->next;
  }
}



int main() {
  SList_t<int> *list_a = sl_list<int>();
  SList_t<int> *list_b = sl_list<int>();

  list_a->pushFront(1);
  list_a->pushFront(0);
  list_a->pushFront(2);

  list_b->pushFront(4);
  list_b->pushFront(8);
  
  list_a->sum_numerals(list_a, list_b,false);
}

File: chapter8/python/chapter8.py

import math
import queue 

class SLNode:
  def __init__(self, value):
    self.val = value
    self.next = None 
    self.child = None 

class SList:
  def __init__(self):
    self.head = None 
  
  @property
  def tail(self):
    runner = self.head
    if (runner == None):
      return runner
    while runner.next:
      runner = runner.next
    return runner

  def pushFront(self, value):
    node = SLNode(value)
    node.next = self.head
    self.head = node
  
  def popFront(self):
    retValue = None 
    if self.head:
      retValue = self.val
    return retValue
  
  def pushBack(self, value):
    node = None
    runner = None  
    if (self.head == None):
      self.pushFront(value)
    else:
      node = SLNode(value)
      runner = self.head 
      while runner.next:
        runner = runner.next 
      runner.next = node 
    
  def contains (self,value):
    runner = this.head 
    while runner:
      if runner.val == value:
        return True
    return False 

  def popBack(self):
    retValue = None 
    runner = self.head
    newTail = None 
    
    if self.head == None :
      return retValue
    
    while runner.next:
      newTail = runner 
      runner = runner.next 
    retValue = runner.val
    
    if runner == self.head:
      self.head = None 
    
    if newTail:
      newTail.next = None 
    
    del runner 
    return retValue

  def removeVal (self, value):
    runner = self.head
    runner_prev = None
    del_node = None   
    rslt = false 
    while (runner):
      if (runner.val == value):
        rslt |= True
        del_node = runner 
        if ( runner == self.head):
          self.head = self.head.next
          runner = self.head
        else:
          runner_prev.next = runner.next
          runner = runner_prev
        del del_node
    runner_prev = runner
    runner = runner.next

  def reverse(self):
    tmp_node  = SLNode(0)
    tmp_node_next = None 
 
    if (self.head == None):
      return None 
    
    while self.head:
      tmp_node.next = self.head 
      self.head = self.head.next
      tmp_node.next.next = tmp_node_next
      tmp_node_next = tmp_node.next
    
    if (tmp_node.next):
      self.head = tmp_node.next
    
    del tmp_node
  
  def isPalindrome(self):
    runner = None 
    start = self.head 
    end = None 

    if start == None:
      return False

    while 1 :
      runner = start 
      while runner.next != end:
        runner = runner.next
      end = runner;  
      if start.val != end.val:
        return False
      if end.next == start or start != end:
        break
      start = start.next
    return True
  
  def kthLastNode(self, k):
    count = 0 
    runner = self.head

    #count 
    while runner:
      runner = runner.next
      count += 1
    
    # count - k
    count -= k 
    if count > 0:
      runner = self.head
      while count > 0:
        runner = runner.next
        count -= 1
    return runner           
  
  def shiftRight(self, n):
    runner = self.head
    newHead = None 
    node = None 
    count = 0

    if n == 0 or runner== None:
      return None 
    
    while runner:
      runnner = runner.next
      count += 1
    
    if n > 0:
      runner = self.head
      for i in range(0, abs(count - n) ):
        runner = runner.next

      #break list 
      newHead = runner.next
      runner.next = None 

      #connect trailing nodes to front 
      runner = newHead
      while runner.next:
        runner = runner.next
      runner.next = self.head
      self.head = newHead

    else:
      return self.shiftLeft(n)
  
  def shiftLeft (self, n):
    runner = self.head
    newHead = None 
    count = 0 

    while runner:
      count += 1
      if count == n:
        #break link 
        newHead = runner.next
        runner.next = None 
        
        #newHead must not be NULL
        if newHead == None:
          break 

        # point runner to tail of new HEAD 
        runner = newHead
        while runner.next:
          runner = runner.next 

        # link new list  with previous  HEAD 
        runner.next = self.head
        
        # update head 
        self.head = newHead  

        break
      runner = runner.next
  
  def sumNumerals (self, lista, listb):
    a = lista.head 
    b = listb.head 
    sum = [0,0, 0] 
    resultList = SList()
    digit = None 
    factor = 1

    # pluck digits into sum array
    while a or b:
      if a:
        sum[0] += a.val * factor
        a = a.next
      if b:
        sum[1] += b.val * factor
        b = b.next
      factor += 10

    #sum accumlators 
    sum[2] = sum[0] + sum[1]
    
    #pluck least sig. digit into list queue  

    while (sum[2] > 0):
      digit = sum[2] % 10
      resultList.pushBack(digit)
      sum[2] = int(sum[2] / 10)
    
    return resultList
  
  def setupLoop (numberOfNodes, loopback_pos):
    myList = SList()
    feedback_node = None 

    for i in range(1, numberOfNodes + 1):
      myList.pushBack(i)
      if loopback_pos == i:
        feedback_node = myList.tail
    
    #feedback
    if feedback_node:
      myList.tail.next = feedback_node

    return myList
  
  def flattenChildren(self):
    runner = None 
    d_queue = queue.Queue()
    children = []
    currTail = None 
    
    if self.head != None :
      d_queue.put(self.head)

    while not d_queue.empty():
      runner = d_queue.get() # pop front of queue 
      while runner:
        if runner.child: # store child node 
          d_queue.put(runner.child)
          static_array_children.append(runner.child)
        runner = runner.next
    
    # append children list to list  (update tail property)
    for child in children:
      self.tail.next = child
      runner = child 
      while runner.next:
        runner = runner.next 
      self.tail = runner 

  def unflattenChildren(self):
    histo = {}
    runner = self.head 
    runner_prev = None 
    tailSet = False 

    while runner:
      if runner.child:
        # hashMap frequency of child node in LL 
        if histo.get(runner.child) == None:
          histo[runner.child] = 0
        histo[runner.child] += 1

        #break if repeat  found 
        if histo[runner.child] == 2:

          if tailSet == False: # set tail  (first unlink) 
            tail = runner_prev
          
          if runner_prev:
            runner_prev.next = None 
          
      runner_prev = runner      
      runner = runner.next

  def hasLoop(self):
    slow = self.head 
    fast = self.head  
    
    while slow:
      if fast:
        fast = fast.next
      if fast:
        fast = fast.next
      slow = slow.next
      if fast == slow:
        return true

  def loopStart(self):
    slow = self.head 
    fast = self.head  
    collision = None 
    loopHead = None 
    runner = None 
    n = 0

    while slow:
      if fast:
        fast = fast.next
      if fast:
        fast = fast.next
      slow = slow.next
      if fast == slow:
        collision = slow 
        break
    
    # n number of nodes in loop
    if collision:
      runner = collision.next 
      while runner != collision:
        n += 1
        runner = runner.next 

      # move loopHead  n-1 nodes from LL Head 
      loopHead = self.head
      for i in range(0, n-1):
        loopHead = loopHead.next 

      #break loop at feeback linking loopHead 
      runner = loopHead
      while runner.next != loopHead:
        runner = runner.next
    
    return runner 
  
  def breakLoop (self):
    feedback = self.loopStart()
    if feedback:
      feedback.next = None 
  
  def numberOfNodes(self):
    feedback = self.loopStart()
    runner = self.head 
    count = 0 

    if runner == None :
      return count
    while runner.next != feedback:
      runner = runner.next 
      count += 1
    return count 
  
  def swapPairs(self):
    runner = self.head 
    runner_prev = None

    while runner:

      if runner_prev:
        
        if self.head == runner_prev:
          self.head = runner
        
        if self.tail == runner:
          self.tail = runner_prev
        
        runner_prev.next = runner.next
        runner.next = runner_prev

        runner = runner_prev
        runner_prev = None 
      
      else:
        runner_prev = runner 
        runner = runner.next


  def display(self):
    runner = self.head 
    while runner:
      print(runner.val, end='')
      runner = runner.next 
    print()



class DLNode:
  def __init__(self, value):
    self.val = value
    self.prev = None 
    self.next = None 

class DList:
  def __init__(self):
    self.head = None
    self.tail = None 

  def push(self, value):
    node = DLNode(value)
    if self.head == None:
      self.head = self.tail = node 
    else:
      self.head.prev = node 
      node.next = self.head 
      self.head = node 

  def pop (self):
    data = None 
    if self.head:
      data = self.head.val
      self.head = self.head.next 
      if self.head:
        del self.head.prev
        self.head.prev = None 
    return data 

  def front (self):
    if (self.head == None):
      return None
    else:
      return self.head.val

  def back (self):
    if (self.tail == None):
      return None 
    else:
      return self.tail.val

  def contains(self, value):
    runner_front = self.head
    runner_back = self.tail
    hasValue = false 
    
    while runner_front or runner_back:
      hasValue |= runner_front.val == value or runner_back.val == value

      if runner_back:
        runner_back = runner_back.prev
      
      if runner_front:
        runner_front = runner_front.next
    
    return hasValue 

  def size(self, value):
    runner_front = self.head
    runner_back = self.tail
    count = 0
    
    while (runner_back or runner_front):
      if runner_back == runner_front: 
        count += 1
      else:
        count += 2
      
      if runner_back:
        runner_back = runner_back.prev
      
      if runner_front:
        runner_front = runner_front.next
    return count 
  
  def prependValue(self, newValue, existingValue):
    runner = self.head
    node = None
    newNode = None  
    while runner:
      if runner.val == existingValue:
        if self.head == runner:
          self.push(newValue)
        else:
          newNode = DLNode(newValue)
          runner.prev.next = newNode
          newNode.prev = runner.prev
          newNode.next = runner
          runner.prev = newNode
      runner = runner.next
  
  def appendValue(self, newValue, existingValue):
    runner = self.head 
    node = None 
    newNode = None 
    while runner:
      if runner.val == existingValue:
        newNode = DLNode(newValue)
        newNode.next = runner.next
        if runner.next:
          runner.next.prev = newNode
        newNode.prev = runner
        runner.next = newNode
      runner = runner.next 
  
  def kthLastNode (self, k):
    runner = self.tail
    data = None 
    
    for i in range(0, k):
      if runner:
        runner = runner.prev
      else:
        return None 
    if runner:
      data = runner.val
    return data
  
  def deleteMiddleNode (self, node):
    node.prev.next = node.next 
    node.next.prev = node.prev 
    del node 

  def isValid(self) :
    slow = None 
    fast = None 

    if self.head:
      slow = self.head
      fast = slow.next 
    else:
      return false

    while fast:

      if fast.prev != slow:
        return False 
      
      if slow:
        slow = slow.next 
      
      if fast:
        fast = fast.next 
    
    return True 

  def partition(self, val):
    node = None 
    runner = self.head 
    
    while runner:
      if (runner.val < val):
        node = runner
        if runner.prev:
          runner.prev.next = runner.next 
        if runner.next:
          runner.next.prev = runner.prev 
        #move node to front 
        node.next = self.head 
        self.head.prev = node 
        self.head = node 
      runner =  runner.next
    
  def isPalindrome(self):
    end = self.tail
    front = self.head 

    while end or front:
      if (end.val != front.val):
        return False 
      if end:
        end = end.prev
      if front:
        front = front.next

      if (end== front or end.next == front):
        break
    return True
  
  def reverse (self):
    back = self.tail
    runner = None 
    if self.tail:
      back = self.tail.prev 
      self.tail.prev = NULL
      self.head = self.tail 
      self.tail = self.head
      runner = self.head 
    while back:
      self.tail = back
      runner.next = back 
      back = back.next 
      runner = runner.next 
      runner.next = None 

  def loopStart(self):
    back = None  
    front = None 

    if self.head == None:
      return None 
      
    back = self.head
    front = self.head.next

    while front:
      
      if front.prev != back:
        break

      if back:
        back = back.next

      if front:
        front = front.next 
    return back

  def breakLoop (self):
    feedback = self.loopStart()
    if feedback:
      feedback.next = None 
  
  def repair(self):
    while not self.isValid:
      self.breakLoop()

  def display(self):
    runner = self.head 
    while  runner:
      print ("[" , runner.val,"]", end='')
      runner = runner.next

dlist = DList()
dlist.push(22)
dlist.push(23)
dlist.appendValue(51, 22)
dlist.display()
